redbot/core/utils/__init__.py

```diff
@@ -255,3 +257,128 @@ def bounded_gather(
+
+    def __aiter__(self) -> AsyncIter[_T]:
+        return self
+
+    async def __anext__(self) -> _T:
+        try:
+            item = next(self._content)
+        except StopIteration:
+            raise StopAsyncIteration
+        self._i += 1
+        if self._i % self._steps == 0:
+            await asyncio.sleep(self._delay)
+        return item
+
+    def __await__(self) -> Generator[Any, None, List[_T]]:
+        return self.flatten().__await__()
+
+    async def flatten(self) -> List[_T]:
+        """Returns a list of the iterable."""
+        return [item async for item in self]
+
+    def filter(self, function: Callable[[_T], Union[bool, Awaitable[bool]]]) -> AsyncFilter[_T]:
+        """
+        Filter the iterable with an (optionally async) predicate.
+
+        Parameters
+        ----------
+        function : Callable[[T], Union[bool, Awaitable[bool]]]
+            A function or coroutine function which takes one item of ``iterable``
+            as an argument, and returns ``True`` or ``False``.
+
+        Returns
+        -------
+        AsyncFilter[T]
+            An object which can either be awaited to yield a list of the filtered
+            items, or can also act as an async iterator to yield items one by one.
+
+        Examples
+        --------
+            >>> from redbot.core.utils import AsyncIter
+            >>> def predicate(value):
+            >>>    return value <= 5
+            >>> iterator = AsyncIter([1, 10, 5, 100])
+            >>> async for i in iterator.filter(predicate):
+            1
+            5
+
+           >>> from redbot.core.utils import AsyncIter
+           >>> def predicate(value):
+           >>>    return value <= 5
+           >>> iterator = AsyncIter([1, 10, 5, 100])
+           >>> await iterator.filter(predicate)
+           [1, 5]
+
+        """
+        return async_filter(function, self)
+
+    def enumerate(self, start: int = 0) -> AsyncIterator[Tuple[int, _T]]:
+        """Async iterable version of `enumerate`.
+
+        Parameters
+        ----------
+        start : int
+            The index to start from. Defaults to 0.
+
+        Returns
+        -------
+        AsyncIterator[Tuple[int, T]]
+            An async iterator of tuples in the form of ``(index, item)``.
+
+        Examples
+        --------
+            >>> from redbot.core.utils import AsyncIter
+            >>> iterator = AsyncIter(['one', 'two', 'three'])
+            >>> async for i in iterator.enumerate(start=10):
+            (10, 'one')
+            (11, 'two')
+            (12, 'three')
+
+        """
+        return async_enumerate(self, start)
+
+    async def without_duplicates(self) -> AsyncIterator[_T]:
+        """
+        Iterates while omitting duplicated entries.
+
+        Examples
+        --------
+            >>> from redbot.core.utils import AsyncIter
+            >>> iterator = AsyncIter([1,2,3,3,4,4,5])
+            >>> async for i in iterator.without_duplicates():
```

Comment:
This looks incomplete, maybe you wanted to add a `print()` here?
```suggestion
            >>> async for i in iterator.without_duplicates():
            >>>     print(i)
```